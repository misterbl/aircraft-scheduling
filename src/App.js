import React, { Component } from "react";
import "./App.scss";
import AirCraftsList from "./components/AirCraftsList";
import FlightsList from "./components/FlightsList";
import TImeline from "./components/Timeline";
import sortArray from "./sortArray";
import { calculatePercentage } from "./utils";
import { TURNAROUND } from "./const";
import planeGif from "./assets/planeGif.gif";
import infoIcon from "./assets/information.svg";
import Instructions from "./components/Instructions";

class App extends Component {
  state = {
    aircraftList: [],
    flightsList: [],
    selectedFlight: null,
    rotation: [],
    flyingTimesList: [],
    loading: false
  };
  componentDidMount() {
    this.getAircrafts();
  }

  getAircrafts = async () => {
    this.setState({ loading: true });
    try {
      const response = await fetch(
        "https://infinite-dawn-93085.herokuapp.com/aircrafts"
      );
      const json = await response.json();
      const aircraftList = json.data;
      this.setState({ loading: false, aircraftList });
    } catch (e) {
      this.setState({
        fatalError: true
      });
    }
  };

  onSelectAircraft = async () => {
    this.setState({ loading: true });
    try {
      const response = await fetch(
        "https://infinite-dawn-93085.herokuapp.com/flights"
      );
      const json = await response.json();
      const result = json.data;
      const flightsList = await sortArray(result, "origin");
      this.setState({ loading: false, flightsList });
    } catch (e) {
      this.setState({
        fatalError: true
      });
    }
  };

  onSelectFlight = async selectedFlight => {
    this.setState({ loading: true });
    const flyingTime =
      selectedFlight.arrivaltime - selectedFlight.departuretime;
    const timePercentage = await calculatePercentage(flyingTime);
    const updatedSelectedFlight = { ...selectedFlight, timePercentage };
    this.setState({ selectedFlight: updatedSelectedFlight });
    await this.addToRotation();
  };

  getTotalAircraftUse = () =>
    Math.round(
      this.state.rotation.reduce(
        (prev, next) => prev + next.timePercentage,
        0
      ) * 100
    ) / 100;

  addToRotation = async () => {
    const { rotation: rotationState, selectedFlight } = this.state;

    const updatedRotation = rotationState.concat(selectedFlight);
    const rotation = sortArray(updatedRotation, "departuretime");

    const isFlightJourneyEligible = await this.checkFlightJourney(rotation);
    const isFlightTimeEligible = await this.checkFlightTime();
    if (
      rotationState.length === 0 ||
      (rotationState.length > 0 &&
        isFlightTimeEligible &&
        isFlightJourneyEligible)
    ) {
      this.setState({ rotation });
      this.getTotalAircraftUse();
      await this.addTurnAround();
      this.removeFromFlightsList();
      this.setState({ loading: false });
    } else {
      alert(
        "this flight doesn't fit this rotation, choose another flight or remove the conflicting flight(s) from the rotation"
      );
      this.setState({ loading: false });
    }
  };
  removeFromFlightsList = async () => {
    this.setState({ loading: false });
    const { flightsList: flightsListState, selectedFlight } = this.state;
    const flightsList = await flightsListState.filter(flight => {
      return flight.id !== selectedFlight.id;
    });
    this.setState({ flightsList });
  };
  prevFlight = (list, index) => list[index - 1] && list[index - 1];

  nextFlight = (list, index) => list[index + 1] && list[index + 1];

  addToFlyingTimeList = async (list, flight) => {
    const updatedList = await list.concat(flight);
    const resultList = await sortArray(updatedList, "departuretime");
    this.setState({ flyingTimesList: resultList });
    return resultList;
  };

  createTurnaround = (departure, arrival) => {
    const flyingTime = departure - arrival;
    const timePercentage = calculatePercentage(flyingTime);
    const turnaround = {
      departuretime: arrival,
      arrivaltime: departure,
      turnaround: true,
      timePercentage
    };
    this.addToFlyingTimeList(this.state.flyingTimesList, turnaround);
  };

  addTurnAround = async () => {
    const {
      selectedFlight,
      flyingTimesList: flyingTimesListFromState
    } = this.state;
    const { departuretime, arrivaltime } = selectedFlight;
    const flyingTimesList = await this.addToFlyingTimeList(
      flyingTimesListFromState,
      selectedFlight
    );
    const selectedFlightIndex = flyingTimesList.indexOf(selectedFlight);
    const prevFlight = this.prevFlight(flyingTimesList, selectedFlightIndex);
    const nextFlight = this.nextFlight(flyingTimesList, selectedFlightIndex);
    const timeBefore = prevFlight && departuretime - prevFlight.arrivaltime;
    const timeAfter =
      selectedFlight && nextFlight && nextFlight.departuretime - arrivaltime;
    if (timeBefore > 0 && prevFlight) {
      this.createTurnaround(departuretime, prevFlight.arrivaltime);
    }
    if (timeAfter > 0 && nextFlight) {
      this.createTurnaround(nextFlight.departuretime, arrivaltime);
    }
  };

  checkFlightTime = () => {
    const {
      rotation,
      selectedFlight: { departuretime, arrivaltime, origin, destination }
    } = this.state;
    return rotation.reduce((carry, flight) => {
      if (
        (flight.destination === origin &&
          departuretime - flight.arrivaltime >= TURNAROUND) ||
        (destination === flight.origin &&
          flight.departuretime - arrivaltime >= TURNAROUND)
      ) {
        return true;
      }
      return carry;
    }, false);
  };

  checkFlightJourney = rotation => {
    let noConflictWithPrev;
    let noConflictWithNext;
    const { selectedFlight } = this.state;
    const flightIndex = rotation.indexOf(selectedFlight);
    const prevFlight = this.prevFlight(rotation, flightIndex);
    const nextFlight = this.nextFlight(rotation, flightIndex);

    if (!prevFlight) {
      noConflictWithPrev = true;
    }
    if (prevFlight && prevFlight.destination === selectedFlight.origin) {
      noConflictWithPrev = true;
    }
    if (!nextFlight) {
      noConflictWithNext = true;
    }
    if (nextFlight && nextFlight.origin === selectedFlight.destination) {
      noConflictWithNext = true;
    }
    return noConflictWithPrev && noConflictWithNext;
  };

  onRemoveFlight = async selectedFlight => {
    this.setState({ loading: true });
    await this.setState({ selectedFlight });
    await this.removeFromRotation();
    this.removeFromFlyingTime();
  };

  removeFromRotation = async () => {
    const {
      rotation: rotationState,
      flightsList: flightsListState,
      selectedFlight
    } = this.state;
    const rotation = rotationState.filter(
      flight => flight.id !== selectedFlight.id
    );
    const updatedFlightList = flightsListState.concat(selectedFlight);
    const flightsList = await sortArray(updatedFlightList, "origin");
    this.setState({ rotation, flightsList });
  };

  removeFromFlyingTime = () => {
    const {
      flyingTimesList: flyingTimesListState,
      selectedFlight
    } = this.state;
    let flyingTimesList;
    const flightIndex = flyingTimesListState.indexOf(selectedFlight);
    const prevFlight = this.prevFlight(flyingTimesListState, flightIndex);
    const nextFlight = this.nextFlight(flyingTimesListState, flightIndex);
    if (prevFlight && prevFlight.turnaround) {
      this.setState(state => {
        flyingTimesList = state.flyingTimesList.filter(
          flight => flight.id !== prevFlight.id
        );
        this.setState({ loading: false });
        return { flyingTimesList };
      });
    }
    if (nextFlight && nextFlight.turnaround) {
      this.setState(state => {
        flyingTimesList = state.flyingTimesList.filter(
          flight => flight.id !== nextFlight.id
        );
        this.setState({ loading: false });
        return { flyingTimesList };
      });
    }
    this.setState(state => {
      flyingTimesList = state.flyingTimesList.filter(
        flight => flight.id !== state.selectedFlight.id
      );
      this.setState({ loading: false });
      return { flyingTimesList };
    });
  };

  render() {
    const {
      aircraftList,
      flightsList,
      rotation,
      flyingTimesList,
      loading
    } = this.state;

    return (
      <div className="text-center">
        {loading && (
          <img
            className="loading-indicator"
            src={planeGif}
            alt="loading indicator"
          />
        )}
        <div className="d-inline-block">
          <h1 className="mt-15 text-center">Aircraft scheduling</h1>
        </div>
        <div className="info-icon">
          <img src={infoIcon} alt="information icon" />
          <Instructions />
        </div>
        <div className="d-flex mt-35 justify-content-center">
          <AirCraftsList
            totalAircraftUse={this.getTotalAircraftUse}
            aircraftList={aircraftList}
            onSelectAircraft={this.onSelectAircraft}
          />
          <div className="d-flex flex-column">
            <FlightsList
              flightsList={rotation}
              rotation
              onRemoveFlight={this.onRemoveFlight}
            />
            {flyingTimesList.length > 0 && (
              <TImeline flyingTimesList={flyingTimesList} />
            )}
          </div>
          <FlightsList
            flightsList={flightsList}
            onSelectFlight={this.onSelectFlight}
            onRemoveFlight={this.onRemoveFlight}
          />
        </div>
      </div>
    );
  }
}

export default App;
