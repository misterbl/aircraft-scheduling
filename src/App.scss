$light-blue-color: #44a3f5;
$jordy-blue: #89c4f4;
$opacity: 0.8;
$small: 15px;
$medium: 30px;
$border-radius: 4px;

body {
  margin: 0;
  padding: 0;
  font-family: "Baloo Chettan", -apple-system, BlinkMacSystemFont, "Segoe UI",
    "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans",
    "Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  background-color: #4b77be;
  color: white;
}

ul {
  list-style-type: none;
  padding: 0px;
}

h3 {
  text-align: center;
}

h1 {
  padding-top: $medium;
  font-family: "Baloo Chettan", sans-serif;
}

.flight {
  border-bottom: 1px solid $jordy-blue;
  &:last-child {
    border-bottom: none;
  }
}

li {
  padding: 5px;
  cursor: pointer;
  margin-bottom: 10px;

  &:hover {
    background-color: $light-blue-color;
    opacity: $opacity;
    border-radius: $border-radius;
  }
}

.container {
  padding: 0px;
  margin: 0px;
  ul {
    border: 1px solid $jordy-blue;
    border-radius: $border-radius;
    margin: $small;
    padding: $small;
    max-height: 500px;
    overflow: scroll;
  }
}

.timeline {
  widows: 864px;
}

.idle-time {
  background-color: grey;
}

.mw-250 {
  max-width: 250px;
}

.mn-250 {
  min-width: 250px;
}
.mt-35 {
  margin-top: 35px;
}

.text-center {
  text-align: center;
}

.bd-black {
  border: 1px solid black;
}
.rotation {
  min-width: 400px;
}

.progress {
  height: 70px;
  margin: $small;
  background-color: #ccc;
}

.instructions {
  display: none;
  max-width: 540px;
  right: 270px;
  padding: $small;
  margin-top: $medium;
  border: 1px $jordy-blue solid;
  border-radius: $border-radius;
  position: absolute;
  background-color: $light-blue-color;
  z-index: 2;
}

.loading-indicator {
  width: 150px;
  position: absolute;
  top: 50%;
  left: 50%;
}

.info-icon {
  cursor: pointer;
  margin: 0 0 29px 10px;
  vertical-align: middle;
  display: inline-block;
  &:hover {
    .instructions {
      display: block;
    }
  }
}
