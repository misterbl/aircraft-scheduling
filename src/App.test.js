import * as React from "react";
import { shallow } from "enzyme";
import App from "./App";

describe("App", () => {
  const props = {
    App: {
      id: "1234",
      origin: "origin",
      readable_departure: "readable_departure",
      destination: "destination",
      readable_arrival: "readable_arrival"
    },
    onRemoveApp: jest.fn(),
    onSelectApp: jest.fn()
  };
  const wrapper = shallow(<App {...props} />);
  it("matches the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  // it("calls onSelectApp if rotation is undefined", () => {
  //   const li = wrapper.find("li");
  //   li.simulate("doubleClick");
  //   expect(props.onSelectApp).toHaveBeenCalled();
  // });
  // it("calls onRemoveApp if rotation is true", () => {
  //   wrapper.setProps({ rotation: true });
  //   const li = wrapper.find("li");
  //   li.simulate("doubleClick");
  //   expect(props.onRemoveApp).toHaveBeenCalled();
  // });
});
