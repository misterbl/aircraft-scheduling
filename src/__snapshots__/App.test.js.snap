// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    App={
      Object {
        "destination": "destination",
        "id": "1234",
        "origin": "origin",
        "readable_arrival": "readable_arrival",
        "readable_departure": "readable_departure",
      }
    }
    onRemoveApp={[MockFunction]}
    onSelectApp={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="mt-15 text-center"
        >
          Aircraft scheduling
        </h1>,
        <Instructions />,
        <div
          className="d-flex mt-35"
        >
          <div
            className="d-flex flex-column"
          >
            <FlightsList
              flightsList={Array []}
              onRemoveFlight={[Function]}
              rotation={true}
            />
            <Timeline
              flyingTimesList={Array []}
            />
          </div>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Aircraft scheduling",
          "className": "mt-15 text-center",
        },
        "ref": null,
        "rendered": "Aircraft scheduling",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            null,
            <div
              className="d-flex flex-column"
            >
              <FlightsList
                flightsList={Array []}
                onRemoveFlight={[Function]}
                rotation={true}
              />
              <Timeline
                flyingTimesList={Array []}
              />
            </div>,
            null,
          ],
          "className": "d-flex mt-35",
        },
        "ref": null,
        "rendered": Array [
          null,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <FlightsList
                  flightsList={Array []}
                  onRemoveFlight={[Function]}
                  rotation={true}
                />,
                <Timeline
                  flyingTimesList={Array []}
                />,
              ],
              "className": "d-flex flex-column",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "flightsList": Array [],
                  "onRemoveFlight": [Function],
                  "rotation": true,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "flyingTimesList": Array [],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
          null,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="mt-15 text-center"
          >
            Aircraft scheduling
          </h1>,
          <Instructions />,
          <div
            className="d-flex mt-35"
          >
            <div
              className="d-flex flex-column"
            >
              <FlightsList
                flightsList={Array []}
                onRemoveFlight={[Function]}
                rotation={true}
              />
              <Timeline
                flyingTimesList={Array []}
              />
            </div>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Aircraft scheduling",
            "className": "mt-15 text-center",
          },
          "ref": null,
          "rendered": "Aircraft scheduling",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              null,
              <div
                className="d-flex flex-column"
              >
                <FlightsList
                  flightsList={Array []}
                  onRemoveFlight={[Function]}
                  rotation={true}
                />
                <Timeline
                  flyingTimesList={Array []}
                />
              </div>,
              null,
            ],
            "className": "d-flex mt-35",
          },
          "ref": null,
          "rendered": Array [
            null,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <FlightsList
                    flightsList={Array []}
                    onRemoveFlight={[Function]}
                    rotation={true}
                  />,
                  <Timeline
                    flyingTimesList={Array []}
                  />,
                ],
                "className": "d-flex flex-column",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "flightsList": Array [],
                    "onRemoveFlight": [Function],
                    "rotation": true,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "flyingTimesList": Array [],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
            null,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
